lib/service.dart

```diff
@@ -192,6 +232,243 @@ class IsolateManager {
+        if (isolate.extensionRPCs != null) {
+          for (String extensionName in isolate.extensionRPCs) {
+            if (extensionName.startsWith('ext.flutter.')) {
+              _setFlutterIsolate(ref);
+              break;
+            }
+          }
+        }
+      }
+      _serviceExtensionManager._addRegisteredExtensionRPCs(ref);
+    }
+  }
+
+  void _setFlutterIsolate(IsolateRef ref) {
+    if (_flutterIsolate == ref) {
+      // Isolate didn't change. Do nothing.
+      return;
+    }
+    _flutterIsolate = ref;
+    _flutterIsolateController.add(ref);
+    _flutterIsolateController.onListen =
+        () => _flutterIsolateController.add(_flutterIsolate);
+  }
+
+  StreamSubscription<IsolateRef> getCurrentFlutterIsolate(Function onData) {
+    return _flutterIsolateController.stream.listen(onData);
+  }
+}
+
+class ServiceExtensionManager {
+  VmService _service;
+  IsolateManager _isolateManager;
+
+  bool firstFrameEventReceived = false;
+
+  final Map<String, StreamController<bool>> _serviceExtensionController =
+      <String, StreamController<bool>>{};
+  final Map<String, StreamController<ServiceExtensionState>>
+      _serviceExtensionStateController =
+      <String, StreamController<ServiceExtensionState>>{};
+
+  /// All available service extensions.
+  final Set<String> _serviceExtensions = Set<String>();
+
+  /// All service extensions that are currently enabled.
+  final Map<String, ServiceExtensionState> _enabledServiceExtensions =
+      <String, ServiceExtensionState>{};
+
+  /// Temporarily stores service extensions that we need to add. We should not add
+  /// extensions until the first frame event has been received [firstFrameEventReceived].
+  final Set<String> _pendingServiceExtensions = Set<String>();
+
+  void _handleExtensionEvent(Event event) {
+    final String extensionKind = event.extensionKind;
+    if (event.kind == 'Extension' && extensionKind == 'Flutter.FirstFrame' ||
+        extensionKind == 'Flutter.Frame') {
+      _onFrameEventReceived();
+    }
+  }
+
+  void _onFrameEventReceived() {
+    if (firstFrameEventReceived) {
+      // The first frame event was already received.
+      return;
+    }
+    firstFrameEventReceived = true;
+
+    _pendingServiceExtensions.forEach(_addServiceExtension);
+    _pendingServiceExtensions.clear();
+  }
+
+  void _addRegisteredExtensionRPCs(IsolateRef isolateRef) async {
+    if (_service == null) {
+      return;
+    }
+    final Isolate isolate = await _service.getIsolate(isolateRef.id);
+    if (isolate.extensionRPCs != null) {
+      isolate.extensionRPCs.forEach(_maybeAddServiceExtension);
+
+      if (!firstFrameEventReceived) {
+        final EvalOnDartLibrary flutterLibrary = new EvalOnDartLibrary(
+          'package:flutter/src/widgets/binding.dart',
+          _service,
+        );
+        flutterLibrary
+            .eval('WidgetsBinding.instance.debugDidSendFirstFrameEvent')
```

Comment:
use dart's async await syntax where possible. It will make some of this code clearer.

Reply:
Done.